File: twoSum.py
'''
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
'''

    ### Naive Solution ###
def twoSumNaive(nums, target):
        for i in range(len(nums)):
            for j in range(len(nums)):
                if i == j: 
                    continue
                else:
                    sum_num = nums[i] + nums[j]        
                if sum_num == target:
                    return [i,j]
                else:
                    continue

   ### Hash Map Solution ###
def twoSumOptmized(nums, target):
        required = {}
        for i in range(len(nums)):
            if target - nums[i] in required:
                return [required[target - nums[i]],i]
            else:
                required[nums[i]]=i

File: twoSum.cpp
#include<iostream>
#include<vector>
#include<map>

std::vector<int> twoSum(std::vector<int> nums, int target) {
    std::map<int, int> required;
    std::vector<int> out;
    for(int i = 0; i <nums.size(); i++){
        if(required.find(target-nums[i]) != required.end()){
            out.push_back(required[target-nums[i]]);
            out.push_back(i);
            return out;
        }
        else{
            required[nums[i]] = i;
        }
    }
return out;
}

int main(){
    int target = 9;
    std::vector<int> nums;
    nums.push_back(2);
    nums.push_back(7);
    nums.push_back(11);
    nums.push_back(15);
    
    std::vector<int> out = twoSum(nums, target);
    for(int i = 0; i<out.size(); i++){
        std::cout << out[i] <<" ";

    }
    return 0;
}